### PR TITLE
feat: delete chat messages

### LINKED ISSUE #350: Delete chat messages
- Creators can delete messages of Mods/Listeners
- Mods can delete listeners chat messages
- Listeners can only delete their own.

I want to work on this if this is not undesired.  @benawad 
Many many people have requested this.

### PR DESCRIPTION
**Pull request by nadirabbas**
Fixes #350 

| Entity | Can delete |
| ------------- | ------------- |
| Room creators  | `own`, `mods`, `listeners` messages |
| Mods  | `own`, `listeners` messages  |
| Listeners | `own` messages  |

Creators and Mods can delete all messages of a particular user at once

| Action | View |
| ------------- | ------------- |
| Delete own message  | image |
| Delete someone else's message  | image |

Working demo:

https://user-images.githubusercontent.com/38838675/109210445-a4654280-77ce-11eb-8b48-7f01169882bf.mp4

### CODE REVIEW

**Comment by nadirabbas** [comment]
@benawad  conflicts resolved

**Comment by benawad** [comment]
1. DogeHouse will be getting a redesign, until then just stick the delete button inside the profile modal instead of adding tooltip +framer motion 
2. This is missing server-side authorization logic. Get the status of each user server-side: https://github.com/benawad/dogehouse/blob/staging/kousa/lib/business-logic/room_logic.ex#L94 and make sure the deleter has higher permission level

**Comment by nadirabbas** [comment]
@benawad  Okay I will add server-side validation. But I didn't understand the first point, how will we delete a particular message from the profile modal?

**Comment by benawad** [comment]
You'll have to add some props to it for message id so you know which one to delete

**Comment by nadirabbas** [comment]
@benawad Please review the changes when you get the time. I also added an updated version of the video above.

File: kofta/package.json [hunk 1]
```diff
@@ -14,6 +14,7 @@
     "@types/react-dom": "^16.9.8",
     "emoji-mart": "^3.0.0",
     "formik": "^2.2.6",
+    "framer-motion": "^3.7.0",
```
Comment (benawad) [inline kofta/package.json]: you can remove the deps

File: kousa/lib/business-logic/room_chat_logic.ex [hunk 2]
```diff
@@ -94,4 +98,38 @@ defmodule Kousa.BL.RoomChat do
...
+    case Kousa.Data.Room.get_room_status(deleter_id) do
+      {:creator, room} ->
+        RegUtils.lookup_and_cast(
+          Gen.RoomChat,
+          room.id,
+          {:message_deleted, deleter_id, message_id}
+        )
+
+      {:mod, room} ->
+        if user_id != room.creatorId do
```
Comment (benawad) [inline kousa/lib/business-logic/room_chat_logic.ex:115]: this allows mods to delete other mods messages. I'm ok with this so we can avoid more DB queries, but can you add a comment about this

File: kousa/lib/gen/room_chat.ex [hunk 3]
```diff
@@ -80,6 +80,48 @@ defmodule Kousa.Gen.RoomChat do
     end
   end
 
+  def handle_cast({:message_deleted, user_id, message_id}, %State{} = state) do
+    last_timestamp = Map.get(state.last_message_map, user_id)
+    {_, seconds, _} = :os.timestamp()
+
+    if not Map.has_key?(state.ban_map, user_id) and
```
Comment (benawad) [inline kousa/lib/gen/room_chat.ex]: I don't think you need to check ban_map and I wouldn't worry about throttling deletes

File: kousa/lib/gen/room_chat.ex [hunk 4]
```diff
@@ -80,6 +80,48 @@ defmodule Kousa.Gen.RoomChat do
...
+    else
+      {:noreply, state}
+    end
+  end
+
+  def handle_cast({:messages_deleted, deleter_id, user_id}, %State{} = state) do
+    last_timestamp = Map.get(state.last_message_map, user_id)
+    {_, seconds, _} = :os.timestamp()
+
+    if not Map.has_key?(state.ban_map, deleter_id) and
```
Comment (benawad) [inline kousa/lib/gen/room_chat.ex]: same as above

**Review by benawad — CHANGES_REQUESTED** [review]
overall, looking really good

**Comment by benawad** [comment]
are you able to run the elixir server locally? It had some small compiler errors when I tried locally that I fixed real quick.

**Comment by nadirabbas** [comment]
@benawad really sorry! I made these changes in the Github editor, figured they are pretty simple 😅

**Comment by nadirabbas** [comment]
what was the mistake?

**Comment by benawad** [comment]
Just a missing `}`, if you use VSCode, install ElixirLS too

**Pull request merged**